arrows/klv/klv_read_write.txx

```diff
@@ -343,6 +356,114 @@ klv_ber_oid_length( T value )
   return ( _int_bit_length( value ) + 6 ) / 7;
 }
 
+// ---------------------------------------------------------------------------
+template < class T, class Iterator >
+double
+klv_read_flint( double minimum, double maximum, Iterator& data, size_t length )
+{
+  KLV_ASSERT_INT( T );
+  KLV_ASSERT_UINT8_ITERATOR( data );
+
+  _check_range_length( minimum, maximum, length );
+
+  // Before read to avoid moving iterator
+  if( std::is_signed< T >::value && minimum != -maximum )
+  {
+    throw std::logic_error( "range must be symmetrical around zero" );
+  }
+
+  auto const int_value = klv_read_int< T >( data, length );
+  auto const float_value = static_cast< double >( int_value );
+
+  // C++17: if constexpr
+  if( std::is_signed< T >::value )
+  {
+    // Special invalid / out-of-range value
+    if( int_value == _int_min< T >( length ) )
+    {
+      return std::numeric_limits< double >::quiet_NaN();
+    }
+
+    auto const scale = maximum / _int_max< T >( length );
+    return float_value * scale;
+  }
+  else
+  {
+    auto const scale = ( maximum - minimum ) / _int_max< T >( length );
+    return float_value * scale + minimum;
+  }
+}
+
+// ---------------------------------------------------------------------------
+template < class T, class Iterator >
+void
+klv_write_flint( double value, double minimum, double maximum, Iterator& data,
+                 size_t length )
+{
+  // Ensure types are compatible with our assumptions
+  KLV_ASSERT_INT( T );
+  KLV_ASSERT_UINT8_ITERATOR( data );
+
+  _check_range_length( minimum, maximum, length );
+
+  // C++17: if constexpr
+  if( std::is_signed< T >::value )
+  {
+    if( minimum != -maximum )
+    {
+      throw std::logic_error( "range must be symmetrical around zero" );
+    }
+
+    auto const max_int = _int_max< T >( length );
+    auto const min_int = _int_min< T >( length );
+    auto const scale = max_int / maximum;
+    auto const float_value = value * scale;
+    auto int_value = static_cast< T >( std::round( float_value ) );
+
+    if( !( value <= maximum && minimum <= value ) )
+    {
+      // Special invalid / out-of-range value
+      int_value = min_int;
+    }
+    else if( float_value >= max_int )
+    {
+      // Ensure floating-point rounding doesn't result in integer overflow
+      int_value = max_int;
+    }
+    else if( float_value <= min_int + 1 )
+    {
+      // Ensure floating-point rounding doesn't result in integer underflow
+      int_value = min_int + 1;
+    }
+    klv_write_int( int_value, data, length );
+  }
+  else
+  {
+    // Clamp out-of-range values
+    value = std::isnan( value ) ? minimum : value;
+    value = std::max( value, minimum );
+    value = std::min( value, maximum );
```

Comment:
This feels probably less than optimally efficient, but I'm not sure it's worth changing, as the more efficient code would be less clear to the reader. However, I might be wrong, as it [looks](https://godbolt.org/z/xxETjor8a) like the compiler may be producing optimal code anyway (my floating-point assembly is inadequate to be *certain*, but...).

However, that said, since you need to guard against overflow from rounding later (and don't early-out on non-finite values), it *might* be more efficient to skip straight to writing `minimum` if given a NaN, and let the later checks handle infinities.